modules/otr.cpp

```diff
@@ -0,0 +1,1291 @@
+        PutModuleContext(ctx, "Authentication aborted.");
+    }
+
+    void CmdGenKey(const CString& sLine) {
+        assert(m_pUserState);
+        const char* accountname = GetUser()->GetUserName().c_str();
+
+        bool bHasKey =
+            otrl_privkey_find(m_pUserState, accountname, PROTOCOL_ID);
+        bool bOverwrite = sLine.Token(1).Equals("--overwrite");
+        bool bReally = sLine.Token(1).Equals("--really");
+
+        if (bHasKey && !bOverwrite) {
+            PutModuleBuffered("Private key already exists. Use " +
+                              Clr(Bold, "genkey --overwrite") +
+                              " to overwrite the old one.");
+            return;
+        }
+
+        if (!bHasKey && !bReally) {
+            PutModuleBuffered(Clr(Red, "WARNING:") +
+                              " This plugin does not provide true end-to-end "
+                              "encryption, as the encryption terminates at "
+                              "the bouncer. You need to make sure that both "
+                              "the bouncer and your client's connection to "
+                              "it are secure.");
+            PutModuleBuffered(Clr(Bold, "NOTE:") +
+                              " If you're running this plugin on a VM, make "
+                              "sure it has sufficient entropy, otherwise ZNC "
+                              "might get stuck. Install haveged to increase "
+                              "available entropy.");
+            PutModuleBuffered(
+                "If you still want to generate new OTR key, type " +
+                Clr(Bold, "genkey --really") + ".");
+            return;
+        }
+
+        if (m_GenKeyRunning) {
+            PutModuleBuffered("Key generation is already running.");
+            return;
+        }
+
+        gcry_error_t err;
+        err = otrl_privkey_generate_start(m_pUserState, accountname,
+                                          PROTOCOL_ID, &m_NewKey);
+        if (err) {
+            PutModuleBuffered(CString("Key generation failed: ") +
+                              gcry_strerror(err));
+            return;
+        }
+
+        PutModuleBuffered("Starting key generation in a background thread.");
+        m_GenKeyRunning = true;
+        AddJob(new COtrGenKeyJob(this));
+    }
+
+    void SaveIgnores() {
+        CString sFlat = "";
+        bool bFirst = true;
+
+        for (VCString::const_iterator it = m_vsIgnored.begin();
+             it != m_vsIgnored.end(); it++) {
+            if (bFirst) {
+                bFirst = false;
+            } else {
+                sFlat += " ";
+            }
+            sFlat += *it;
+        }
+        SetNV("ignore", sFlat, true);
+    }
+
+    bool IsIgnored(const CString& sNick) {
+        CString sNickLower = sNick.AsLower();
+
+        for (VCString::const_iterator it = m_vsIgnored.begin();
+             it != m_vsIgnored.end(); it++) {
```

Comment:
Oh, cool, didn't know C++11 had this. I can't use this syntax when deleting elements from the vector, right?